src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectOptionsProcessor.cs

```diff
@@ -0,0 +1,182 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.IO;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal class VisualStudioProjectOptionsProcessor : IDisposable
+    {
+        private readonly VisualStudioProject _project;
+        private readonly HostWorkspaceServices _workspaceServices;
+        private readonly ICommandLineParserService _commandLineParserService;
+
+        /// <summary>
+        /// Gate to guard all mutable fields in this class.
+        /// The lock hierarchy means you are allowed to call out of this class and into <see cref="_project"/> while holding the lock.
+        /// </summary>
+        private object _gate = new object();
+        private string _commandLine = "";
+        private CommandLineArguments _commandLineArgumentsForCommandLine;
+        private string _explicitRuleSetFilePath;
+        private IReferenceCountedDisposable<IRuleSetFile> _ruleSetFile = null;
+
+        public VisualStudioProjectOptionsProcessor(VisualStudioProject project, HostWorkspaceServices workspaceServices)
+        {
+            _project = project ?? throw new ArgumentNullException(nameof(project));
+            _workspaceServices = workspaceServices;
+            _commandLineParserService = workspaceServices.GetLanguageServices(project.Language).GetRequiredService<ICommandLineParserService>();
+
+            // Set up _commandLineArgumentsForCommandLine to a default. No lock taken since we're in the constructor so nothing can race.
+            ReparseCommandLine_NoLock();
+        }
+
+        public string CommandLine
+        {
+            get
+            {
+                return _commandLine;
+            }
+
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                lock (_gate)
+                {
+                    if (_commandLine == value)
+                    {
+                        return;
+                    }
+
+                    _commandLine = value;
+                }
+
+                ReparseCommandLine_NoLock();
```

Comment:
>ReparseCommandLine_NoLock [](start = 16, length = 25)

Are these supposed to be outside of the lock?

Reply:
Hah!